lldb/tools/lldb-rpc/lldb-rpc-gen/server/RPCServerSourceEmitter.cpp

```diff
@@ -0,0 +1,592 @@
+  uint8_t ArgIdx = 0;
+  for (QualType ParamType : FuncProtoType->param_types()) {
+    Param CallbackParam;
+    CallbackParam.IsFollowedByLen = false;
+    CallbackParam.Type = ParamType;
+    if (ParamType->isVoidPointerType())
+      CallbackParam.Name = "baton";
+    else
+      CallbackParam.Name = "arg" + std::to_string(ArgIdx++);
+
+    CallbackParams.push_back(CallbackParam);
+    CallbackParamsAsStrings.push_back(ParamType.getAsString(method.Policy) +
+                                      " " + CallbackParam.Name);
+  }
+  const std::string CallbackReturnTypeName =
+      FuncProtoType->getReturnType().getAsString(method.Policy);
+  const std::string CallbackName = method.MangledName + "_callback";
+
+  // Emit Function Header
+  std::string Header;
+  llvm::raw_string_ostream HeaderStream(Header);
+  HeaderStream << "static " << CallbackReturnTypeName << " " << CallbackName
+               << "(" << llvm::join(CallbackParamsAsStrings, ", ") << ") {";
+  EmitLine(Header);
+  IndentLevel++;
+
+  // Emit Function Body
+  EmitLine("// RPC connection setup and sanity checking");
+  EmitLine("CallbackInfo *callback_info = (CallbackInfo *)baton;");
+  EmitLine("rpc_common::ConnectionSP connection_sp = "
+           "rpc_common::Connection::GetConnectionFromID(callback_info->"
+           "connection_id);");
+  EmitLine("if (!connection_sp)");
+  IndentLevel++;
+  if (FuncProtoType->getReturnType()->isVoidType())
+    EmitLine("return;");
+  else
+    EmitLine("return {};");
+  IndentLevel--;
+
+  EmitLine("// Preparing to make the call");
+  EmitLine("static RPCFunctionInfo g_func(\"" + CallbackName + "\");");
+  EmitLine("RPCStream send;");
+  EmitLine("RPCStream response;");
+  EmitLine("g_func.Encode(send);");
+
+  EmitLine("// The first thing we encode is the callback address so that the "
+           "client-side can know where the callback is");
+  EmitLine("RPCValueEncoder(send, rpc_common::RPCPacket::ValueType::Argument, "
+           "callback_info->callback);");
+  EmitLine("// Encode all the arguments");
+  for (const Param &CallbackParam : CallbackParams) {
+    if (lldb_rpc_gen::TypeIsSBClass(CallbackParam.Type)) {
+
+      // FIXME: SB class server references are stored as non-const references so
+      // that we can actually change them as needed. If a parameter is marked
+      // const, we will fail to compile because we cannot make an
+      // SBFooServerReference from a `const SBFoo &`.
+      // To work around this issue, we'll apply a `const_cast` if needed so we
+      // can continue to generate callbacks for now, but we really should
+      // rethink the way we store object IDs server-side to support
+      // const-qualified parameters.
+      QualType UnderlyingSBClass =
+          lldb_rpc_gen::GetUnderlyingType(CallbackParam.Type);
+      QualType UnqualifiedUnderlyingSBClass =
+          UnderlyingSBClass.getUnqualifiedType();
+
+      // FIXME: Replace this bespoke logic with a nice function in RPCCommon.
```

Comment:
Does RPCCommon actually exist or is that also a future thing?

Either way doesn't sound critical, just wondering.

Reply:
It's in a separate PR that Chelsea created. There's a circular dependency between this PR and that PR unfortunately.